Conversion of Stabilizer to canonical CliffordOperator (resolving #64)

adrianariton opened the pull request:
The main thing I thought would be a first try to make the implementation work is just to store the operations as they are performed (swapping and pivoting) and undo them. There is deffinetly a better way of doing this by building the inverse matrix to undo the gaussian elimination iteratively and I will try to implement it later this week if this is the way of doing it.

My only issue right now is that I think modifying the `canonicalize_gott!` return argument messes it's other calls (from what I see it is used multiple times, in `QuantumClifford.jl` and `graphs.jl`). 

Also, I see that by the current return of `canonicalize_gott!`, only the x and z permutations get returned for them to be inverted later, but not also the `mul_left` operations. To also invert the `mul_left` operations, my first thought was to add this vector of operations.

I also rewrote the `undoperm` in MixedDestabilizer(Stabilizer), and on a few test I randomly selected and inputed manually it seems to work!


- ~~**Edit**: Will fix: Used `s` instead of `r + s`. [QuantumClifford.jl, lines 749-779 undoperm]~~ `DONE`
- **Edit 2**: Phases get a bit scrambled, probably from commutating row-operations. I will fix this soon by changing mul_left (lines
768-769) to it's inverse! `DONE` `and also commented it, because I don't know if it is necesarry to reverse multiplication ops`

adrianariton commented:
I noticed that most tests fail when also undoing the `mul_left` operation, maybe has something to do with the fact that it brings the lines to the original form, which `canonicalize!` does not (it makes them linearly independent). So I commented that part for now.

Krastanov commented on src/canonicalization.jl:
```diff
@@ -218,42 +218,77 @@ See also: [`canonicalize!`](@ref), [`canonicalize_rref!`](@ref)
+    # print(xperm, '\n', zperm, '\n')
+    # There must be a cleaner implementation by iteratively creating the
+    # inverse of the gaussian elimination matrix
+    stabilizer, r, s, xperm, zperm, ops
```
We can not afford to modify the return type of this function as this is a public function on which people depend.

Krastanov replied:
Maybe an appropriate solution would be to add a new keyword argument `recordops::Val{Bool}=Val(false)` and condition the extra information on it.

Krastanov commented on src/canonicalization.jl:
```diff
@@ -218,42 +218,77 @@ See also: [`canonicalize!`](@ref), [`canonicalize_rref!`](@ref)
     rows, columns = size(stabilizer)
+    ops = []
+    retxperm = collect(1:rows)
+    retzperm = collect(1:rows)
```
retxperm and retzperm do not seem to be used for anything

Krastanov commented on src/QuantumClifford.jl:
```diff
@@ -747,7 +747,45 @@ function MixedDestabilizer(stab::Stabilizer{T}; undoperm=true) where {T}
+        for op in reverse(ops)
+            # print("\n\nPerforming: ", op, '\n')
+            if op[1] == 1
+                # show(IOContext(stdout::IO, :limit => true), t)   
```
If something as complicated as this is needed for printing (as opposed to `@show` or `println`) there is an issue that should be reported in the bug tracker. Printing should be easy.

adrianariton replied:
I wanted to use this to be able to check bigger outputs with diff, to see where they differ!

Krastanov commented on src/canonicalization.jl:
```diff
@@ -218,42 +218,77 @@ See also: [`canonicalize!`](@ref), [`canonicalize_rref!`](@ref)
+    push!(ops, (4, 0, 0))
+    # print("\n-------------\n")
+    # print('\n', ops, '\n')
+    # 1-swap, 2-mul, 3-permx, 4-permz
```
I am not a big fan of having this implicit integer encoding. Given that these operations are all in sequence, I think I would prefer if we use 4 separate arrays.

adrianariton replied:
You're right, I will modify this immediately. Also from what I notice on tests, I don't think it is necesary to also track the multiply_left operations (even though the Stabiliser part of the MixedDestab looks like the original stab and that is nice, canonicalise! seems to want the lines linearly-independent).

adrianariton replied:
I returned and undid all operations: row swap, aswell as mul_left. Should I be alarmed if the tests from test_stabs do not work, or do I also need to redo them given the new modified output for MixedDestabilizer?

adrianariton commented:
I added a new commit, in which I cleaned the code and also added the `trackoperations` option to MixedDestabilizer, defaulted on false, that tracks and undoes row operations: 

```Julia
# QuantumClifford.jl , line 714
function MixedDestabilizer(stab::Stabilizer{T}; undoperm=true, trackoperations=false) where {T}
...
```
This is then used in the conversion of Stab to CliffordOp:
```Julia
# dense_cliffords.jl, line 65
function CliffordOperator(stab::Stabilizer)
    md = MixedDestabilizer(copy(stab); trackoperations=true)
    return CliffordOperator(tab(md))
end
```

Krastanov commented:
A new file for the tests would be best. I am trying to follow the SciML style of tests where test files are small and selfcontained.

My bad on the email, I will respond shortly.

adrianariton commented:
Ok! I will start working on the test file later this weekend.

adrianariton commented:
Hello! I added a bunch of tests, checking the conversion went down correctly!
I do think there are some more tests I will be needing to add, could you help me with a range of operations I should check between the stab and it's conversion?

Krastanov commented on src/canonicalization.jl:
```diff
@@ -215,45 +215,62 @@ Based on [gottesman1997stabilizer](@cite).
     permute!(stabilizer,zperm)
-    stabilizer, r, s, xperm, zperm
+
+    if R
```
You are allocating a lot of expensive arrays even if `R` is false. It would be better if they do not get populated at all. Each `push!` can be prefaced by `R && push!(...)`. That way they get executed only if R is true.

also, there is a bunch new whitespace/newlines added. Could you clean it up?

Krastanov commented on test/test_stabtocliff.jl:
```diff
@@ -0,0 +1,36 @@
+        stab = random_stabilizer(4)
+        cl = CliffordOperator(stab)
+        cltab = tab(cl)
+        @test Stabilizer(cltab[end÷2+1:end]) == (stab);
```
could you clean up the extra parentheses and semicolons?

Krastanov commented on test/test_stabtocliff.jl:
```diff
@@ -0,0 +1,36 @@
+
+    for i in 1:10
+
+        gnd4 = S"ZIII
```
you can create this with `one(Stabilizer, ...)`

Krastanov commented on src/QuantumClifford.jl:
```diff
@@ -711,9 +711,10 @@ mutable struct MixedDestabilizer{T<:Tableau} <: AbstractStabilizer
 
 # Added a lot of type assertions to help Julia infer types
-function MixedDestabilizer(stab::Stabilizer{T}; undoperm=true) where {T}
+function MixedDestabilizer(stab::Stabilizer{T}; undoperm=true, trackoperations=false) where {T}
```
May we call it `undorowops` instead of `trackoperations`? Add an explanation of what it is to the docstring.

Krastanov commented on src/canonicalization.jl:
```diff
@@ -215,45 +215,62 @@ Based on [gottesman1997stabilizer](@cite).
 """
-function canonicalize_gott!(stabilizer::Stabilizer; phases::Bool=true)
-    @valbooldispatch _canonicalize_gott!(stabilizer; phases=Val(phases)) phases
+function canonicalize_gott!(stabilizer::Stabilizer; phases::Bool=true, recordops::Bool=false)
```
the `recordops` needs to be explained in the docstring

Krastanov commented on test/test_stabtocliff.jl:
```diff
@@ -0,0 +1,36 @@
+        stab = random_stabilizer(4)
+        cl = CliffordOperator(stab)
+        cltab = tab(cl)
+        @test apply!(gnd4, cl) == (stab);
```
I am very surprised that this works without the need for canonicalization. Something might be wrong.

adrianariton replied:
Shouldn't the Clifford operator bring the ground state to the desired one? Why do we need an extra canonicalize?

Krastanov commented:
A few conflicts exist now, so it would take some work to get this in a mergeable condition again. I will close it for now.

Thank you for tackling this, it is greatly appreciated! If you feel like taking on it again, feel free to ping me.
